### ISSUE TITLE
Use the Azure OpenAI and Azure AI Search built-in connectors in automated Standard logic app workflows

### ISSUE THREAD

**Issue by hallongley** [hallongley/EA-2023#1]
4/4/25: Develop Azure Service Apps and Resource Groups

**Comment by hallongley** [comment]
Azure Cloud Integration with LLM models
 How does an Azure Cloud application integrate with LLMs?Integrating an Azure Cloud application with Large Language Models (LLMs) like DeepSeek or others involves leveraging Azure's infrastructure, services, and tools to host, manage, and interact with the LLM while ensuring scalability, security, and performance. Below, I’ll break down the process step-by-step, focusing on how Azure facilitates this integration.


1. Understanding the Components

Azure Cloud: Microsoft Azure provides a suite of cloud services, including compute (e.g., Azure Virtual Machines, Azure Kubernetes Service), storage (e.g., Azure Blob Storage), networking, and AI/ML services (e.g., Azure Machine Learning, Azure Cognitive Services).
LLM (e.g., DeepSeek): A large language model is a machine learning model designed for natural language processing (NLP) tasks like text generation, classification, translation, or question answering. DeepSeek, for instance, is an open-source LLM optimized for research and applications.
Integration Goal: The Azure application will interact with the LLM to process user inputs, generate responses, and potentially fine-tune or deploy the model for specific use cases.


2. Steps for IntegrationStep 1: Model Hosting and Deployment

Choose a Hosting Option:
Azure Machine Learning (AML): Use AML to deploy the LLM as a managed endpoint. AML supports custom models, so you can upload a pre-trained DeepSeek model (e.g., in ONNX, PyTorch, or TensorFlow format) and deploy it on Azure's infrastructure.
Azure Kubernetes Service (AKS): For more control, deploy the LLM on AKS, which is ideal for scaling GPU-intensive workloads (LLMs often require GPUs like NVIDIA A100 for inference).
Azure Container Instances (ACI): For simpler deployments, use ACI to run the LLM in a containerized environment.
Containerization:
Package the LLM into a Docker container with its dependencies (e.g., Python, PyTorch, Transformers library).
Push the container to Azure Container Registry (ACR) for deployment.
GPU Support:
LLMs like DeepSeek are computationally intensive. Use Azure's GPU-enabled VMs (e.g., NC-series or ND-series) to handle inference efficiently.
Deployment:
Deploy the model as a REST API endpoint using AML or AKS. Azure provides auto-scaling and load balancing to handle varying traffic.


Step 2: Application Integration

Azure App Service or Azure Functions:
Build your application using Azure App Service (for web apps) or Azure Functions (for serverless architectures).
The application will send requests to the LLM's endpoint (e.g., via HTTP POST requests) with user inputs (e.g., text prompts) and receive generated responses.
API Communication:
The LLM endpoint exposes a REST API. Your application can use libraries like requests in Python to send prompts and retrieve responses.
Example:pythonimport requests  endpoint = "[https://<your-llm-endpoint>.azurewebsites.net/api/inference"](https://%3Cyour-llm-endpoint%3E.azurewebsites.net/api/inference) headers = {"Authorization": f"Bearer {api_key}"} payload = {"prompt": "What is the capital of France?"} response = requests.post(endpoint, json=payload, headers=headers) print(response.json()) # Output: {"response": "The capital of France is Paris."}
Authentication:
Secure the LLM endpoint using Azure Active Directory (AAD) or API keys to ensure only authorized applications can access it.


Step 3: Data Storage and Management

Azure Blob Storage:
Store training data, fine-tuning datasets, or model artifacts in Azure Blob Storage. For example, if you need to fine-tune DeepSeek for a specific domain (e.g., legal or medical), you can store the dataset here.
Azure Data Lake:
For large-scale data processing, use Azure Data Lake to manage and preprocess datasets before feeding them to the LLM.
Azure Cosmos DB:
Store user interactions, prompts, and responses in Azure Cosmos DB for low-latency access and analytics.


Step 4: Fine-Tuning and Optimization (Optional)

Fine-Tuning:
Use Azure Machine Learning to fine-tune the LLM on domain-specific data. AML supports distributed training with frameworks like PyTorch or Hugging Face Transformers.
Example: Fine-tune DeepSeek on a dataset of customer support queries to improve its performance in a chatbot application.
Optimization:
Use tools like ONNX Runtime or Azure's model optimization features to reduce inference latency and memory usage.
Apply techniques like quantization or pruning to make the LLM more efficient for deployment.


Step 5: Monitoring and Scaling

Azure Monitor:
Monitor the LLM endpoint for performance metrics (e.g., latency, throughput) and errors using Azure Monitor.
Application Insights:
Integrate Application Insights into your Azure application to track user interactions and LLM performance.
Auto-Scaling:
Configure auto-scaling in AKS or AML to handle spikes in traffic. For example, if your application sees increased usage, Azure can automatically provision more resources to maintain performance.


Step 6: Security and Compliance

Azure Security Center:
Use Azure Security Center to protect your application and LLM endpoint from threats.
Data Privacy:
Ensure compliance with regulations like GDPR or HIPAA by using Azure's encryption and access control features.
Store sensitive data (e.g., user prompts) in encrypted form using Azure Key Vault for key management.
Network Security:
Deploy the LLM endpoint in a Virtual Network (VNet) to restrict access and reduce exposure to the public internet.


Step 7: User Interaction and Feedback Loop

Front-End Integration:
Build a front-end (e.g., using Azure Static Web Apps) to allow users to interact with the LLM-powered application.
Example: A chatbot interface where users type queries, and the app sends them to the LLM for processing.
Feedback Loop:
Collect user feedback on LLM responses and store it in Azure Cosmos DB.
Use this feedback to periodically retrain or fine-tune the model for better accuracy.


3. Example Architecture:

          Here’s a high-level architecture for an Azure Cloud application integrated with an LLM like DeepSeek:

User: Interacts with the application via a web or mobile interface.
Azure App Service/Azure Functions: Hosts the application logic and communicates with the LLM endpoint.
Azure Machine Learning/AKS: Hosts the DeepSeek LLM as a REST API endpoint, running on GPU-enabled infrastructure.
Azure Blob Storage: Stores datasets, model weights, and logs.
Azure Cosmos DB: Stores user prompts, responses, and feedback.
Azure Monitor/Application Insights: Monitors performance and usage.
Azure Active Directory: Manages authentication and access control.


4. Challenges and Considerations

Cost Management:
LLMs are resource-intensive. GPU-enabled VMs can be expensive, so optimize usage with auto-scaling and spot instances where possible.
Latency:
Inference with LLMs can be slow. Use techniques like batching requests or caching common responses (e.g., with Azure Redis Cache) to improve performance.
Model Size:
DeepSeek and similar LLMs can be large (e.g., 7B+ parameters). Ensure your Azure infrastructure has enough memory and compute power.
Ethical Use:
Monitor the LLM's outputs for bias, toxicity, or misinformation. Implement content filters or human-in-the-loop moderation if needed.


5. Example Use Case: Chatbot for Customer Support

Scenario: A company builds a customer support chatbot using DeepSeek on Azure.
Implementation:
Deploy DeepSeek on AKS with GPU support.
Build a serverless Azure Functions app to handle user queries and send them to the LLM endpoint.
Store conversation history in Azure Cosmos DB.
Use Azure Monitor to track chatbot performance and user satisfaction.
Result: The chatbot can answer customer queries in natural language, escalate complex issues to human agents, and improve over time with fine-tuning.


6. Alternative: Using Azure Cognitive ServicesIf you don’t want to manage a custom LLM like DeepSeek, Azure offers pre-built AI models via Azure Cognitive Services (e.g., Azure OpenAI Service, which provides access to models like GPT-3 or GPT-4). These services are easier to integrate but may not offer the same flexibility as a custom model like DeepSeek.



Integration:
Use the Azure OpenAI Service API to send prompts and receive responses.
Example:
pythonfrom azure.ai.textanalytics 
import TextAnalyticsClient from azure.core.credentials 
import AzureKeyCredential  key = "<your-key>" endpoint = "<your-endpoint>" client = TextAnalyticsClient(endpoint=endpoint, credential=AzureKeyCredential(key)) response = client.analyze_sentiment(documents=["What is the capital of France?"]) print(response)


Pros: Fully managed, easy to use, and optimized for Azure.
Cons: Less control over the model, and you’re limited to the models Microsoft provides.


Conclusion



Integrating an Azure Cloud application with an LLM like DeepSeek involves deploying the model on Azure's infrastructure (e.g., AML or AKS), building an application to interact with the model via APIs, and leveraging Azure's storage, monitoring, and security services to ensure a robust solution. 



The process can be tailored to your specific needs, whether you’re building a chatbot, a content generation tool, or a research application. If you’d like to dive deeper into any specific step (e.g., fine-tuning or cost optimization), let me know! 



 Post-Training: Large-Scale Reinforcement Learning on the Base Model 



DeepSeek 

 

We directly apply reinforcement learning (RL) to the base model without relying on supervised fine-tuning (SFT) as a preliminary step. This approach allows the model to explore chain-of-thought (CoT) for solving complex problems, resulting in the development of DeepSeek-R1-Zero. DeepSeek-R1-Zero demonstrates capabilities such as self-verification, reflection, and generating long CoTs, marking a significant milestone for the research community. Notably, it is the first open research to validate that reasoning capabilities of LLMs can be incentivized purely through RL, without the need for SFT. This breakthrough paves the way for future advancements in this area.
We introduce our pipeline to develop DeepSeek-R1. The pipeline incorporates two RL stages aimed at discovering improved reasoning patterns and aligning with human preferences, as well as two SFT stages that serve as the seed for the model's reasoning and non-reasoning capabilities. We believe the pipeline will benefit the industry by creating better models.


Azure AI Integration models:

 

Use the Azure OpenAI and Azure AI Search built-in connectors in automated Standard logic app workflows. These connectors support multiple authentication types, such as API keys, Microsoft Entra ID, and managed identities.
Build an end-to-end Generative AI workflow using Azure OpenAI, AzureML and Snowflake.
Deploy an Azure OpenAI model.
Build a Prompt Flow.
Connect Snowflake data to AzureML.
Coordinate the flow with an AzureML Notebook.
Add rich AI capabilities to an Azure Database for PostgreSQL flexible server instance using the azure_ai extension. It covers integrating both Azure OpenAI and the Azure AI Language service into your database using the extension.


Azure and Snowflake storage integration:

 

To connect Snowflake to Azure, you can use either a Storage Integration or a Shared Access Signature (SAS) Token. Here are the steps for both methods:



Using Storage Integration

Create a Cloud Storage Integration in Snowflake: CREATE STORAGE INTEGRATION azure_int TYPE = EXTERNAL_STAGE STORAGE_PROVIDER = 'AZURE' ENABLED = TRUE AZURE_TENANT_ID = '<tenant_id>' STORAGE_ALLOWED_LOCATIONS = ('azure://<account>.blob.core.windows.net/<container>/<path>/');
Grant Snowflake Access to the Storage Locations: DESC STORAGE INTEGRATION azure_int; Navigate to the URL in the AZURE_CONSENT_URL column and grant permissions.
Create an External Stage: CREATE STAGE my_azure_stage STORAGE_INTEGRATION = azure_int URL = 'azure://myaccount.blob.core.windows.net/container1/path1' FILE_FORMAT = my_csv_format;


Using SAS Token

Generate the SAS Token: Log into the Azure portal. Navigate to your storage account and generate a SAS token with appropriate permissions.
Create an External Stage: CREATE OR REPLACE STAGE my_azure_stage URL='azure://myaccount.blob.core.windows.net/mycontainer/load/files' CREDENTIALS=(AZURE_SAS_TOKEN='?sv=20[1](https://www.bing.com/ck/a?!&&p=3259fe2df91086712a749ced457b590e4613453a17ccead1d002e7a7f21e783fJmltdHM9MTc0MzU1MjAwMA&ptn=3&ver=2&hsh=4&fclid=25db2065-076e-6015-3269-329d06806127&u=a1aHR0cHM6Ly9kb2NzLnNub3dmbGFrZS5jb20vZW4vdXNlci1ndWlkZS9kYXRhLWxvYWQtYXp1cmUtY29uZmln&ntb=1)6-05-31&ss=b&srt=sco&sp=rwdl&se=[2](https://www.bing.com/ck/a?!&&p=96ce35687da8c1fb2b1bf4d3f7e36a0df24589b077a032a967885d639e491f2bJmltdHM9MTc0MzU1MjAwMA&ptn=3&ver=2&hsh=4&fclid=25db2065-076e-6015-3269-329d06806127&u=a1aHR0cHM6Ly9kb2NzLnNub3dmbGFrZS5jb20vZW4vdXNlci1ndWlkZS9wcml2YXRlbGluay1henVyZQ&ntb=1)018-06-27T10:05:50Z&st=2017-06-27T02:05:50Z&spr=https,http&sig=bgqQwoXwxzuD2GJfagRg7VOS8hzNr3QLT7rhS8OFRLQ%3D') FILE_FORMAT = my_csv_format;


Important Considerations

Permissions: Ensure that you have the necessary permissions in both Snowflake and Azure to perform these operations[1](https://www.bing.com/ck/a?!&&p=3259fe2df91086712a749ced457b590e4613453a17ccead1d002e7a7f21e783fJmltdHM9MTc0MzU1MjAwMA&ptn=3&ver=2&hsh=4&fclid=25db2065-076e-6015-3269-329d06806127&u=a1aHR0cHM6Ly9kb2NzLnNub3dmbGFrZS5jb20vZW4vdXNlci1ndWlkZS9kYXRhLWxvYWQtYXp1cmUtY29uZmln&ntb=1)2.
Security: Use secure methods for storing and managing credentials, such as using storage integrations instead of directly embedding credentials in SQL commands[1](https://www.bing.com/ck/a?!&&p=3259fe2df91086712a749ced457b590e4613453a17ccead1d002e7a7f21e783fJmltdHM9MTc0MzU1MjAwMA&ptn=3&ver=2&hsh=4&fclid=25db2065-076e-6015-3269-329d06806127&u=a1aHR0cHM6Ly9kb2NzLnNub3dmbGFrZS5jb20vZW4vdXNlci1ndWlkZS9kYXRhLWxvYWQtYXp1cmUtY29uZmln&ntb=1).
Performance: Configure network routing to use Microsoft network routing for improved query performance1.


By following these steps, you can securely connect Snowflake to Azure and manage your data efficiently.
